[AIRFLOW-1998] Implemented DatabricksRunNowOperator for jobs/run-now …

isknight opened the pull request:
### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title.
  - https://issues.apache.org/jira/browse/AIRFLOW-1998

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

  - Added a new Databrick's operator, DatabricksRunNowOperator, for the API's jobs/run-now functionality.
  - After consulting Databricks on their preference, pulled shared code out of DatabricksSubmitRunOperator into module level functions that both the DatabricksSubmitRunOperator & DatabricksRunNowOperator call.
  - Added required API call to databricks_hook.py
  - Added/updated tests for both databricks_operators.py and the databricks_hook.py
  - Fixed a connection state issue with tokens in the setUp part of the operator tests where conn.extras wasn't getting reset between test runs and maintaining its mock token.
  - Fixed some existing indentation for flake8

### Tests

- [x] My PR adds the following unit tests:

- I mirrored the same tests that the existing DatabricksSubmitRunOperator had for the new DatabricksRunNowOperator.

  - **Added**:
tests/contrib/operators/test_databricks_operator.py
	* DatabricksOperatorSharedFunctions
		* test_deep_string_coerce
	* DatabricksRunNowOperatorTest
		* test_init_with_named_parameters
		* test_init_with_json
		* test_init_with_merging
		* test_init_with_templating
		* test_init_with_bad_type
		* test_exec_success
		* test_exec_failure
		* test_on_kill
  - **Updated**:
tests/contrib/operators/test_databricks_operator.py
	* DatabricksSubmitRunOperatorTest
		* test_init_with_named_parameters
		* test_init_with_json
		* test_init_with_merging
		* test_init_with_templating
		* test_init_with_bad_type
		* test_exec_success
		* test_exec_failure
		* test_on_kill
  - **Removed**:
tests/contrib/operators/test_databricks_operator.py
	* DatabricksSubmitRunOperatorTest
		* test_deep_string_coerce (it was moved to its own external test class)
		
		
### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

riot-isknight committed:
[AIRFLOW-1998] Implemented DatabricksRunNowOperator for jobs/run-now endpoint.

andrewmchen commented on airflow/contrib/hooks/databricks_hook.py:
```diff
@@ -143,6 +144,18 @@ def _do_api_call(self, endpoint_info, json):
+        """
+        Utility function to call the ``api/2.0/jobs/run-now`` endpoint.
+
+        :param json: The data used in the body of the request to the ``submit`` endpoint.
```
nit: `submit` -> `run-now`

isknight replied:
Oops! nice catch. I read through my new docs like 5 times but managed to miss it. :) I'll update.

andrewmchen commented on airflow/contrib/operators/databricks_operator.py:
```diff
@@ -30,6 +30,81 @@
+        raise AirflowException(msg)
+
+
+def _handle_databricks_operator_execution(operator, context):
```
Can we explicitly pass the parameters: `hook` and `log`, into this function and lift the logic to initially invoke `run-now` or `run-submit` back into the operator? The end result would be a helper function responsible for polling for the run result.

With this solution, we'd duplicate a bit of logic with the xcom business (which we can refactor later) but we avoid having to pass the wide interface of an `operator` (which is really a special databricks operato) to this function.

andrewmchen commented on airflow/contrib/operators/databricks_operator.py:
```diff
@@ -30,6 +30,81 @@
+            time.sleep(operator.polling_period_seconds)
+
+
+def _handle_databricks_operator_on_kill(operator):
```
For this one, I'm thinking we can simply duplicate the logic in both operators.

isknight replied:
Yeah, I felt iffy when I pulled that out. It's not good to duplicate code, but what if their lifecycle deviates? I'll update this.

andrewmchen reviewed (changes requested):
Looks very good overall. I just had a couple comments about passing the "operator" to the module functions.

riot-isknight committed:
Per feedback: fixed a documentation error

isknight committed:
Merge pull request #1 from apache/master

Pulling in head fork changes

riot-isknight committed:
Per feedback, reintegrated the execute and on_kill onto the objects.

riot-isknight committed:
- Addressed remaining feedback on new DatabricksRunNowOperator
- Merge branch 'master' into Airflow-1998-PR

# Conflicts:
#	tests/contrib/hooks/test_databricks_hook.py
#	tests/contrib/operators/test_databricks_operator.py

isknight committed:
Merge pull request #2 from apache/master

Merging latest upstream into my fork

riot-isknight committed:
- Merged upstream into my fork.
- fixed a minor flake8 issue

Merge branch 'master' into Airflow-1998-PR

# Conflicts:
#	airflow/contrib/hooks/databricks_hook.py

riot-isknight committed:
Reverting two files (squash incoming)

andrewmchen commented on airflow/contrib/operators/databricks_operator.py:
```diff
@@ -30,6 +30,66 @@
+
+def _handle_databricks_operator_execution(operator, hook, log, context):
+    """
+    Handles the Airflow + Databricks lifecycle logic for a data bricks operator
```
nit: `Databricks`

riot-isknight committed:
Fixed a  documentation issue.
